lib/translator/write.cc

```diff
@@ -0,0 +1,217 @@
+                       uint32_t nsid, uint32_t transfer_length_pages) {
+  StatusCode status_code = allocation.SetPages(transfer_length_pages, 0);
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+
+  nvme_wrapper.cmd = {.opc = static_cast<uint8_t>(nvme::NvmOpcode::kWrite),
+                      .psdt = 0,  // prps are used
+                      .nsid = nsid};
+
+  nvme_wrapper.cmd.dptr.prp.prp1 = allocation.data_addr;
+
+  return status_code;
+}
+
+// Builds NVMe cdw 12 for Write10, Write12, Write16 translations
+uint32_t BuildCdw12(uint16_t transfer_length, uint8_t prinfo, bool fua) {
+  // cdw12 nlb bits 15:00 (zero based field), prinfo bits 29:26, fua bit 30
+  return (static_cast<uint32_t>(fua) << 30) |
+         (static_cast<uint32_t>(prinfo) << 26) | (transfer_length - 1);
+}
+
+StatusCode Write(bool fua, uint8_t wrprotect, uint32_t transfer_length,
+                 NvmeCmdWrapper& nvme_wrapper, Allocation& allocation,
+                 uint32_t nsid, uint32_t page_size, uint32_t lba_size) {
+  if (transfer_length == 0) {
+    DebugLog("NVMe read command does not support transfering zero blocks\n");
+    return StatusCode::kNoTranslation;
+  }
+
+  transfer_length &= 0xffff;  // truncate to 16 bits
+  StatusCode status_code =
+      LegacyWrite(nvme_wrapper, allocation, nsid,
+                  GetTransferLengthPages(transfer_length, page_size, lba_size));
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+
+  uint8_t pr_info;
+  status_code = BuildPRInfo(wrprotect, pr_info);
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+  nvme_wrapper.cmd.cdw[2] = htoll(BuildCdw12(transfer_length, pr_info, fua));
+
+  return status_code;
+}
+
+}  // namespace
+
+StatusCode Write6ToNvme(Span<const uint8_t> scsi_cmd,
+                        NvmeCmdWrapper& nvme_wrapper, Allocation& allocation,
+                        uint32_t nsid, uint32_t page_size, uint32_t lba_size) {
+  scsi::Write6Command write_cmd = {};
+  if (!ReadValue(scsi_cmd, write_cmd)) {
+    DebugLog("Malformed Write6 Command");
+    return StatusCode::kInvalidInput;
+  }
+
+  // A TRANSFER LENGTH field set to zero specifies that 256 logical blocks shall
+  // be written. Any other value specifies the number of logical blocks that
+  // shall be written (Section 3.59) of
+  // https://www.seagate.com/files/staticfiles/support/docs/manual/Interface%20manuals/100293068j.pdf
+  uint16_t updated_transfer_length =
+      (write_cmd.transfer_length == 0) ? 256 : write_cmd.transfer_length;
+  StatusCode status_code = LegacyWrite(
+      nvme_wrapper, allocation, nsid,
+      GetTransferLengthPages(updated_transfer_length, page_size, lba_size));
+
+  if (status_code != StatusCode::kSuccess) {
+    return status_code;
+  }
+
+  // cdw10 Starting lba bits 31:00
+  uint32_t host_endian_lba = (write_cmd.logical_block_address_1 << 16) +
+                             ntohs(write_cmd.logical_block_address_2);
```

Comment:
will this ever overflow? would using `|` be appropriate?

Reply:
This wouldn't overflow because `logical_block_address_1` is a `uint8_t` (although im not too sure how the implicit cast works here). and I think the `|` ends up being the same as `+`

Reply:
`logical_block_address_1` is 5 bits and `logical_block_address_2` is 16 bits. So, they should both fit into a 32 bit integer. 

https://github.com/googleinterns/scsi2nvme/blob/master/lib/scsi.h#L691